### PR TITLE
Add two levels to use_c10_dispatcher

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26434 Move Generator ops to c10
* #26273 Enable creation of boxing wrappers for some aten operators
* **#26272 Add two levels to use_c10_dispatcher**
* #26337 c10::KernelFunction

```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml
Comment (dzhulgakov) [inline aten/src/ATen/native/native_functions.yaml]: nit: I don't think you need quotes in yaml

**Review by dzhulgakov — APPROVED** [review]
It looks ok to me

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: This gonna need to do the multiple dispatch thing.

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: Also going to need multiple dispatch bits too.

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: Can we add an `assert['use_c10_dispatcher'] == 'full'` here? (or whatever that value is)

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: You don't need parentheses here

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: assert here plz

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: assert here

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: assert here

**Review by ezyang — APPROVED** [review]
There is nothing obviously wrong with this patch, but this patch adds **four** more blocks of generated code to `function_wrapper.py`. As I have commented in the patch earlier in this stack, I think at least one of these is unnecessary: the basic API should work whether or not the boxing code works or not. So you shouldn't have to switch between these cases in the C10 dispatcher. When you want to enforce boxing works on everything, you just remove the sfinae guard in C10 side.

File: aten/src/ATen/function_wrapper.py
Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: ouch I think this got screwed up by a merge conflict with the revert of the multi dispatch diff.

File: aten/src/ATen/function_wrapper.py
Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: this is using `inferred_type_Set`, same as the aten version, so should be ok, right?

Ultimately we want to move this logic into the dispatcher, but I'm waiting for more of your dispatch diffs to land before doing that because it would likely conflict.

File: aten/src/ATen/function_wrapper.py
Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: There is an assertion in `native_parse.yaml` but I'll also add one here.

File: aten/src/ATen/function_wrapper.py
Comment (smessmer) [reply aten/src/ATen/function_wrapper.py]: nope

**Comment by smessmer** [comment]
@ezyang saw your comments after landing it. Created a new PR for addressing them: https://github.com/pytorch/pytorch/pull/26587

File: aten/src/ATen/function_wrapper.py
Comment (ezyang) [reply aten/src/ATen/function_wrapper.py]: Multidispatch is landed, so all of the stuff you need is landed already.

**Pull request closed without merging**
